crates/factor-llm/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+mod host;
+
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use spin_factors::{
+    ConfigureAppContext, Factor, InstanceBuilders, PrepareContext, RuntimeFactors,
+    SelfInstanceBuilder,
+};
+use spin_locked_app::MetadataKey;
+use spin_world::v1::llm::{self as v1};
+use spin_world::v2::llm::{self as v2};
+
+pub const ALLOWED_MODELS_KEY: MetadataKey<Vec<String>> = MetadataKey::new("ai_models");
+
+pub struct LlmFactor {
+    create_engine: Box<dyn Fn() -> Box<dyn LlmEngine> + Send + Sync>,
+}
+
+impl LlmFactor {
+    pub fn new<F>(create_engine: F) -> Self
+    where
+        F: Fn() -> Box<dyn LlmEngine> + Send + Sync + 'static,
+    {
+        Self {
+            create_engine: Box::new(create_engine),
+        }
+    }
+}
+
+impl Factor for LlmFactor {
+    type RuntimeConfig = ();
+    type AppState = AppState;
+    type InstanceBuilder = LlmDispatch;
+
+    fn init<T: RuntimeFactors>(
+        &mut self,
+        mut ctx: spin_factors::InitContext<T, Self>,
+    ) -> anyhow::Result<()> {
+        ctx.link_bindings(spin_world::v1::llm::add_to_linker)?;
+        ctx.link_bindings(spin_world::v2::llm::add_to_linker)?;
+        Ok(())
+    }
+
+    fn configure_app<T: RuntimeFactors>(
+        &self,
+        ctx: ConfigureAppContext<T, Self>,
+    ) -> anyhow::Result<Self::AppState> {
+        let component_allowed_models = ctx
+            .app()
+            .components()
+            .map(|component| {
+                Ok((
+                    component.id().to_string(),
+                    component
+                        .get_metadata(ALLOWED_MODELS_KEY)?
+                        .unwrap_or_default()
+                        .into_iter()
+                        .collect::<HashSet<_>>()
+                        .into(),
+                ))
+            })
+            .collect::<anyhow::Result<_>>()?;
+        Ok(AppState {
+            component_allowed_models,
+        })
+    }
+
+    fn prepare<T: RuntimeFactors>(
+        &self,
+        ctx: PrepareContext<Self>,
+        _builders: &mut InstanceBuilders<T>,
+    ) -> anyhow::Result<Self::InstanceBuilder> {
+        let allowed_models = ctx
+            .app_state()
+            .component_allowed_models
+            .get(ctx.app_component().id())
+            .cloned()
+            .unwrap_or_default();
+
+        Ok(LlmDispatch {
+            engine: (self.create_engine)(),
```

Comment:
Is `create_engine` supposed to be cheap? This gets called unconditionally whether a component uses llm or not.

Reply:
It's relatively cheap for both implementations (some cloning of some data structures happening), but there's no I/O happening in either.